File: e-rtmp-lab/wip/e-flv-tool/src/flv/parser.ts
import { open } from 'node:fs/promises';

export interface FlvHeader {
  signature: string;   // 'FLV'
  version: number;
  hasAudio: boolean;
  hasVideo: boolean;
  dataOffset: number;
}

export interface FlvTag {
  tagType: number;     // 8 = audio, 9 = video, 18 = script data
  dataSize: number;
  timestamp: number;
  streamId: number;
  data: Uint8Array;
}

export interface FlvFile {
  header: FlvHeader;
  tags: FlvTag[];
}

export interface ParseFlvFileOptions {
  includeData?: boolean;
}

function parseFlvHeader(buf: Uint8Array): FlvHeader {
  if (buf.length < 9) {
    throw new Error('Buffer too small for FLV header (need 9 bytes)');
  }

  const signature = String.fromCharCode(buf[0], buf[1], buf[2]);
  if (signature !== 'FLV') {
    throw new Error(`Invalid FLV signature: '${signature}'`);
  }

  const version = buf[3];
  const flags = buf[4];
  const hasAudio = (flags & 0x04) !== 0;
  const hasVideo = (flags & 0x01) !== 0;
  const dataOffset = (buf[5] << 24) | (buf[6] << 16) | (buf[7] << 8) | buf[8];

  return { signature, version, hasAudio, hasVideo, dataOffset };
}

function readUint24BE(buf: Uint8Array, offset: number): number {
  return (buf[offset] << 16) | (buf[offset + 1] << 8) | buf[offset + 2];
}

/**
 * Parse FLV from an in-memory buffer.
 */
export function parseFlv(buffer: Uint8Array): FlvFile {
  const header = parseFlvHeader(buffer);

  const tags: FlvTag[] = [];
  let offset = header.dataOffset;

  // Skip PreviousTagSize0
  offset += 4;

  while (offset + 11 <= buffer.length) {
    const tagType = buffer[offset];
    const dataSize = readUint24BE(buffer, offset + 1);
    const timestampLow = readUint24BE(buffer, offset + 4);
    const timestampExt = buffer[offset + 7];
    const timestamp = (timestampExt << 24) | timestampLow;
    const streamId = readUint24BE(buffer, offset + 8);

    offset += 11;

    if (offset + dataSize > buffer.length) {
      break; // Truncated tag
    }

    const data = buffer.slice(offset, offset + dataSize);
    tags.push({ tagType, dataSize, timestamp, streamId, data });

    offset += dataSize;

    // Skip PreviousTagSize
    offset += 4;
  }

  return { header, tags };
}

/**
 * Parse FLV from a file using sequential reads (constant memory for headers-only mode).
 */
export async function parseFlvFile(path: string, options?: ParseFlvFileOptions): Promise<FlvFile> {
  const includeData = options?.includeData ?? false;
  let fh;

  try {
    fh = await open(path, 'r');
  } catch (err: unknown) {
    if (err instanceof Error && 'code' in err && (err as NodeJS.ErrnoException).code === 'ENOENT') {
      throw new Error(`File not found: '${path}'`);
    }
    throw new Error(`Could not open file '${path}'`);
  }

  try {
    // Read FLV header (9 bytes)
    const headerBuf = new Uint8Array(9);
    const { bytesRead: headerBytes } = await fh.read(headerBuf, 0, 9, null);
    if (headerBytes < 9) {
      throw new Error('File too small for FLV header');
    }
    const header = parseFlvHeader(headerBuf);

    // Skip to end of header (in case dataOffset > 9) + PreviousTagSize0 (4 bytes)
    const skipBytes = header.dataOffset - 9 + 4;
    if (skipBytes > 0) {
      const skipBuf = new Uint8Array(skipBytes);
      await fh.read(skipBuf, 0, skipBytes, null);
    }

    // Read tags
    const tags: FlvTag[] = [];
    const tagHeaderBuf = new Uint8Array(11);
    const prevTagSizeBuf = new Uint8Array(4);

    while (true) {
      const { bytesRead } = await fh.read(tagHeaderBuf, 0, 11, null);
      if (bytesRead === 0) break; // EOF
      if (bytesRead < 11) break;  // Truncated

      const tagType = tagHeaderBuf[0];
      const dataSize = readUint24BE(tagHeaderBuf, 1);
      const timestampLow = readUint24BE(tagHeaderBuf, 4);
      const timestampExt = tagHeaderBuf[7];
      const timestamp = (timestampExt << 24) | timestampLow;
      const streamId = readUint24BE(tagHeaderBuf, 8);

      let data: Uint8Array;
      if (includeData) {
        data = new Uint8Array(dataSize);
        await fh.read(data, 0, dataSize, null);
      } else {
        // Skip over tag body
        data = new Uint8Array(0);
        if (dataSize > 0) {
          const skipData = new Uint8Array(Math.min(dataSize, 65536));
          let remaining = dataSize;
          while (remaining > 0) {
            const chunk = Math.min(remaining, skipData.length);
            await fh.read(skipData, 0, chunk, null);
            remaining -= chunk;
          }
        }
      }

      tags.push({ tagType, dataSize, timestamp, streamId, data });

      // Read PreviousTagSize
      await fh.read(prevTagSizeBuf, 0, 4, null);
    }

    return { header, tags };
  } finally {
    await fh.close();
  }
}

export function mergeFlvFiles(
  _a: FlvFile,
  _b: FlvFile,
  _options?: { multitrack?: boolean }
): Uint8Array {
  // TODO: Implement FLV merge logic
  //   - If multitrack: interleave tracks using E-FLV multitrack extensions
  //   - Otherwise: concatenate tags, re-timestamp file B after file A
  //   - Write new FLV header + merged tags + PreviousTagSize entries

  return new Uint8Array(0);
}

File: e-rtmp-lab/wip/e-flv-tool/flv/parser.go
package flv

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"os"
)

// TagType represents the type of an FLV tag.
type TagType byte

const (
	TagTypeAudio  TagType = 8
	TagTypeVideo  TagType = 9
	TagTypeScript TagType = 18
)

// FLVHeader represents the 9-byte FLV file header.
type FLVHeader struct {
	Signature  [3]byte
	Version    uint8
	HasAudio   bool
	HasVideo   bool
	DataOffset uint32
}

func parseHeader(f *os.File) (FLVHeader, error) {
	var buf [9]byte
	if _, err := f.Read(buf[:]); err != nil {
		return FLVHeader{}, fmt.Errorf("reading header: %w", err)
	}

	sig := [3]byte{buf[0], buf[1], buf[2]}
	if sig != [3]byte{'F', 'L', 'V'} {
		return FLVHeader{}, fmt.Errorf("invalid FLV signature: %q", sig)
	}

	return FLVHeader{
		Signature:  sig,
		Version:    buf[3],
		HasAudio:   buf[4]&0x04 != 0,
		HasVideo:   buf[4]&0x01 != 0,
		DataOffset: binary.BigEndian.Uint32(buf[5:9]),
	}, nil
}

// DumpFLV reads an FLV/E-FLV file and prints structural information.
func DumpFLV(inputPath string, jsonOutput bool, verbose bool) error {
	f, err := os.Open(inputPath)
	if err != nil {
		return fmt.Errorf("opening file: %w", err)
	}
	defer f.Close()

	info, err := f.Stat()
	if err != nil {
		return fmt.Errorf("stat file: %w", err)
	}

	header, err := parseHeader(f)
	if err != nil {
		return err
	}

	fmt.Printf("File: %s\n", inputPath)
	fmt.Printf("Size: %d bytes\n", info.Size())
	fmt.Printf("Format: FLV/E-FLV\n")
	fmt.Println()
	fmt.Printf("Header\n")
	fmt.Printf("  Version:     %d\n", header.Version)
	fmt.Printf("  Has Audio:   %t\n", header.HasAudio)
	fmt.Printf("  Has Video:   %t\n", header.HasVideo)
	fmt.Printf("  Data Offset: %d\n", header.DataOffset)

	if _, err := f.Seek(int64(header.DataOffset), io.SeekStart); err != nil {
		return fmt.Errorf("seek to data offset: %w", err)
	}

	r := bufio.NewReaderSize(f, 1<<20)

	var previousTagSize [4]byte
	if _, err := io.ReadFull(r, previousTagSize[:]); err != nil {
		if err == io.EOF {
			fmt.Println()
			fmt.Printf("Tags\n")
			fmt.Printf("  Total:  0\n")
			fmt.Printf("  Audio:  0\n")
			fmt.Printf("  Video:  0\n")
			fmt.Printf("  Script: 0\n")
			fmt.Printf("  Other:  0\n")
			return nil
		}
		return fmt.Errorf("reading first previous tag size: %w", err)
	}

	var totalTags uint64
	var audioTags uint64
	var videoTags uint64
	var scriptTags uint64
	var otherTags uint64
	var metadataBlocks [][]AMF0Property
	var codecConfigs []CodecConfig
	var tagHeader [11]byte
	for {
		_, err := io.ReadFull(r, tagHeader[:])
		if err == io.EOF {
			break
		}
		if err != nil {
			if err == io.ErrUnexpectedEOF {
				return fmt.Errorf("reading tag header: truncated file")
			}
			return fmt.Errorf("reading tag header: %w", err)
		}

		totalTags++
		tagType := TagType(tagHeader[0] & 0x1f)
		dataSize := int64(tagHeader[1])<<16 | int64(tagHeader[2])<<8 | int64(tagHeader[3])

		switch tagType {
		case TagTypeVideo:
			videoTags++
			cfg, err := tryParseVideoConfig(r, int(dataSize))
			if err != nil {
				return fmt.Errorf("reading video tag payload: %w", err)
			}
			if cfg != nil {
				codecConfigs = append(codecConfigs, *cfg)
			}
			goto readPrevTagSize
		case TagTypeAudio:
			audioTags++
			cfg, err := tryParseAudioConfig(r, int(dataSize))
			if err != nil {
				return fmt.Errorf("reading audio tag payload: %w", err)
			}
			if cfg != nil {
				codecConfigs = append(codecConfigs, *cfg)
			}
			goto readPrevTagSize
		case TagTypeScript:
			scriptTags++
			props, err := parseScriptTag(r, int(dataSize))
			if err != nil {
				return fmt.Errorf("reading script tag payload: %w", err)
			}
			if props != nil {
				metadataBlocks = append(metadataBlocks, props)
			}
			goto readPrevTagSize
		default:
			otherTags++
		}

		if _, err := io.CopyN(io.Discard, r, dataSize); err != nil {
			if err == io.EOF || err == io.ErrUnexpectedEOF {
				return fmt.Errorf("reading tag payload: truncated file")
			}
			return fmt.Errorf("reading tag payload: %w", err)
		}

	readPrevTagSize:
		if _, err := io.ReadFull(r, previousTagSize[:]); err != nil {
			if err == io.EOF || err == io.ErrUnexpectedEOF {
				return fmt.Errorf("reading previous tag size: truncated file")
			}
			return fmt.Errorf("reading previous tag size: %w", err)
		}
	}

	fmt.Println()
	fmt.Printf("Tags\n")
	fmt.Printf("  Total:  %d\n", totalTags)
	fmt.Printf("  Audio:  %d\n", audioTags)
	fmt.Printf("  Video:  %d\n", videoTags)
	fmt.Printf("  Script: %d\n", scriptTags)
	fmt.Printf("  Other:  %d\n", otherTags)

	for i, props := range metadataBlocks {
		fmt.Println()
		if len(metadataBlocks) == 1 {
			fmt.Printf("onMetaData\n")
		} else {
			fmt.Printf("onMetaData #%d\n", i+1)
		}
		for _, p := range props {
			printAMF0Property(p, 1)
		}
	}

	for _, cfg := range codecConfigs {
		fmt.Println()
		printCodecConfig(cfg)
	}

	// TODO: Parse tag headers for timestamps, stream IDs, data sizes
	// TODO: Detect and report E-FLV track information
	// TODO: Output as text or JSON based on jsonOutput flag
	// TODO: Include offsets, timestamps, tag counts when verbose is true

	return nil
}

// parseScriptTag reads dataSize bytes from r and, if the first AMF0 value is
// the string "onMetaData", returns the properties of the second AMF0 value.
// Returns nil properties (no error) if this is not an onMetaData tag.
func parseScriptTag(r io.Reader, dataSize int) ([]AMF0Property, error) {
	payload := make([]byte, dataSize)
	if _, err := io.ReadFull(r, payload); err != nil {
		return nil, err
	}

	// First AMF0 value should be a string.
	name, offset, err := parseAMF0Value(payload, 0)
	if err != nil {
		return nil, nil // not parseable, skip
	}

	nameStr, ok := name.(string)
	if !ok || nameStr != "onMetaData" {
		return nil, nil
	}

	// Second AMF0 value should be an object or ECMA array.
	value, _, err := parseAMF0Value(payload, offset)
	if err != nil {
		return nil, nil
	}

	props, ok := value.([]AMF0Property)
	if !ok {
		return nil, nil
	}
	return props, nil
}

func printAMF0Property(p AMF0Property, indent int) {
	prefix := ""
	for i := 0; i < indent; i++ {
		prefix += "  "
	}
	switch v := p.Value.(type) {
	case []AMF0Property:
		fmt.Printf("%s%s:\n", prefix, p.Name)
		for _, sub := range v {
			printAMF0Property(sub, indent+1)
		}
	case float64:
		n := int64(v)
		if v == float64(n) {
			if (p.Name == "videocodecid" || p.Name == "audiocodecid") && n > 15 {
				// Values 0–15 are legacy CodecId's. Values > 15 are a FourCC from E-RTMP.
				fourCC := [4]byte{byte(n >> 24), byte(n >> 16), byte(n >> 8), byte(n)}
				fmt.Printf("%s%s: %d (%s)\n", prefix, p.Name, n, string(fourCC[:]))
			} else {
				fmt.Printf("%s%s: %d\n", prefix, p.Name, n)
			}
		} else {
			fmt.Printf("%s%s: %g\n", prefix, p.Name, v)
		}
	case bool:
		fmt.Printf("%s%s: %t\n", prefix, p.Name, v)
	case string:
		fmt.Printf("%s%s: %s\n", prefix, p.Name, v)
	case nil:
		fmt.Printf("%s%s: null\n", prefix, p.Name)
	default:
		fmt.Printf("%s%s: %v\n", prefix, p.Name, v)
	}
}
